### DISCUSSION TITLE
Add option to enable querystring in asset URLs

### DISCUSSION THREAD

**Discussion by slavafyi** [barrel/shopify-vite#96] [Ideas]
Hi. I'm developing a solution to help reuse javascript modules in shopify theme more easily with import-maps. For example:

```js
import { trapFocus } from 'theme.js'`
```
instead of 
```js
import { trapFocus } from 'https://store.myshopify.com/cdn/shop/t/13/assets/theme.e355d783.min.js'
```

So is there any chance that we can add an option to enable/disable [this](https://github.com/barrel/shopify-vite/pull/21) feature? This can be enabled by default, but it would be cool to be able to disable it. Because hashes are difficult to work with if we don't use the build process. For example, a developer who works with a theme directly from the store admin.

**Comment by montalvomiguelo** [comment]
Hi @slavamak (:

If we enable the version parameter. How would ES modules be imported at their latest version from generated chunks? _They don't know the version parameter generated via_ `asset_url`/`Liquid`.

**Reply by slavafyi** [reply]
We will use import-maps. It would look like this:

```liquid
<script type="importmap">
{
  "imports": {
    "customers.js": "{{ 'customers.js' | asset_url }}",
    "theme.js": "{{ 'theme.js' | asset_url }}",
    "localization-form.js": "{{ 'localization-form.js' | asset_url }}",
    "theme.css": "{{ 'theme.css' | asset_url }}",
    "localization-form.css": "{{ 'localization-form.css' | asset_url }}"
  }
}
</script>
```

You can check it in my playground store [here](https://shop.slvmk.dev/) (password `yuussaa`).

I modify `vite/modulepreload-polyfill` so that it uses our generated import-maps instead of `import.meta.url`. The map will also contain css files, although this is not supported in the spec, but we need it to get the correct url with version parameter when loading modules dynamically. For example like `localization-form` in my store above.

I'm not sure how pretty it looks, but it works anyway 😁

**Reply by montalvomiguelo** [reply]
Thanks a lot for sharing @slavamak, that is so cool 🙏 

1. How would we handle the generation of the import-maps, for any number of entries?
2. Would we need to adjust the import-maps during dev and prod?
3. How would we handle the tweaks for vite/modulepreload-polyfill?

**Reply by slavafyi** [reply]
Yeah I like this idea too, although it has some "problems", I'll write about them below 😁

1. Basically we will generate import-maps after write the bundle, i.e. the map will include all chunks and assets file.
2. I guess we need import-maps only for prod. Because during dev the vite server handles all imports itself.
3. I just copied polyfill from vite source code and changed it to make it work with import-maps :)

I see a couple of "problems" here. One problem is that import-maps only support js files, but our map will also contain css files. This doesn't cause errors, but it doesn't follow the [specification](https://github.com/WICG/import-maps).

The second problem is that vite generates imports in chunks with relative paths starts with "./", for some reason this doesn't work with import-maps (and we get 404), so I just replace all "./" with "". Not sure if this is the right way, but it works.

I made a quick version of the plugin, you can check it [here](https://github.com/slavamak/vite-plugin-shopify-import-maps). This is just an idea, would love to hear your thoughts on it :)

**Reply by montalvomiguelo** [reply]
Fantastic job on your plugin! 🚀 

Prolly css urls can be consumed from a separate map to keep the import-maps aligned to the specification.?
The approach for 2nd problem looks good to me; alternatively,[ module specifier keys ](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/script/type/importmap#paths_in_the_module_specifier_map_key)can contain `./` [`base`](https://github.com/barrel/shopify-vite/blob/main/packages/vite-plugin-shopify/src/config.ts#L32)

**Reply by slavafyi** [reply]
Thanks Miguel! In fact in the spec I didn't find mention something about css files, so I guess it's not a problem, but I'm not 100% sure. I added a `<base>` tag and also added `import.meta.resolve` which automatically resolves the url for module specifire and solves the second problem. It seems to work like a charm :) but we need to test this, also requires polyfill for this, but I think it's not a problem

**Reply by montalvomiguelo** [reply]
@slavamak I saw those updates and the playground. Very cool stuff! 🚀
The CSS piece is tricky, AFIK import-maps are primarily designed for managing JS module dependencies.

**Reply by slavafyi** [reply]
@montalvomiguelo Yes right, and it's weird, yet it also works with css and images. Maybe it's a bug? 😁

**Reply by montalvomiguelo** [reply]
@slavamak  I'll add the flexibility of enabling the version parameter through the `versionNumbers` option.  The value of your plugin will be very beneficial in the Shopify theme dev space (: 
cc @wturnerharris

**Reply by montalvomiguelo** [reply]
@slavamak @wturnerharris, I added the `versionNumbers` option in #99 (:

**Comment by jonathanmoore** [comment]
I would love to see @slavamak's idea expanded and an option we could use.  From a premium theme's perspective I would be 99% certain Shopify's review process would flag and decline any theme submission that breaks the native asset_url filter and Shopify's use of cache variables.

**Reply by slavafyi** [reply]
This is already being used in the theme store, so I guess breaks the native asset_url filter not a problem :) You can check Paper theme and [toolkit](https://github.com/BrickspaceLab/paper-toolkit)

**Reply by tylerhellner** [reply]
I would love to use @slavamak's plugin as well! Keep us posted if you're ever thinking about integrating it with this setup.
